### PR TITLE
Added rmw_client, rmw_destroy_client and rmw_service_server_is_available tests

### PR DESCRIPTION
**Pull request by ahcorde**
Testing:
 - rmw_client
 - rmw_destroy_client
 - rmw_service_server_is_available

Related to this PR
   - https://github.com/ros2/rmw_fastrtps/pull/447

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
Related with this PR https://github.com/ros2/rmw_fastrtps/pull/447

File: test_rmw_implementation/test/test_client.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  EXPECT_FALSE(is_available) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  rmw_service_t * service = rmw_create_service(
+    node, ts, "service_name_test",
+    &rmw_qos_profile_default);
+  ASSERT_NE(nullptr, client) << rcutils_get_error_string().str;
+  ret = rmw_service_server_is_available(node, client, &is_available);
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:102]: @ahcorde this one's flaky -- you may want to use one of the testing macros to wait for it.

File: test_rmw_implementation/test/test_client.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+  ret = rmw_service_server_is_available(node, nullptr, &is_available);
+  EXPECT_EQ(ret, RMW_RET_ERROR) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  ret = rmw_service_server_is_available(node, client, &is_available);
+  EXPECT_EQ(ret, RMW_RET_OK) << rmw_get_error_string().str;
+  EXPECT_FALSE(is_available) << rmw_get_error_string().str;
+  rmw_reset_error();
+
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:95]: @ahcorde nit: it'd be nice to separate "bad argument" tests from "good argument" tests.

**Review by hidmic — COMMENTED** [review]
There's one overlap with https://github.com/ros2/rmw_implementation/pull/138 we have to sort out.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12481/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7448/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10194/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12400/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12484/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7451/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10197/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12403/)

File: test_rmw_implementation/test/test_client.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+TEST_F(
+  CLASSNAME(TestServiceServerIsAvailable, RMW_IMPLEMENTATION),
+  rmw_service_server_is_available_bad_args)
+{
+  bool is_available;
+  rmw_ret_t ret = rmw_service_server_is_available(nullptr, nullptr, &is_available);
+  EXPECT_EQ(ret, RMW_RET_ERROR) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  ret = rmw_service_server_is_available(node, nullptr, &is_available);
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:90]: @ahcorde what if the third argument is `nullptr`?

File: test_rmw_implementation/test/test_client.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  EXPECT_EQ(ret, RMW_RET_OK) << rmw_get_error_string().str;
+  EXPECT_TRUE(is_available) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  ret = rmw_destroy_service(node, service);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+TEST_F(CLASSNAME(TestServiceServerIsAvailable, RMW_IMPLEMENTATION), rmw_client_destroy_bad_args)
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:131]: @ahcorde I think this one is fully covered by #138.

File: test_rmw_implementation/test/test_client.cpp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  ret = rmw_destroy_client(node, nullptr);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  ret = rmw_destroy_client(node, client_to_destroy);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestServiceServerIsAvailable, RMW_IMPLEMENTATION), rmw_create_client_fault)
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:151]: @ahcorde this one's good to keep though.

File: test_rmw_implementation/test/test_client.cpp [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  rmw_service_server_is_available_bad_args)
+{
+  bool is_available;
+  rmw_ret_t ret = rmw_service_server_is_available(nullptr, nullptr, &is_available);
+  EXPECT_EQ(ret, RMW_RET_ERROR) << rmw_get_error_string().str;
+  rmw_reset_error();
+
+  ret = rmw_service_server_is_available(node, nullptr, &is_available);
+  EXPECT_EQ(ret, RMW_RET_ERROR) << rmw_get_error_string().str;
+  rmw_reset_error();
```
Comment (hidmic) [inline test_rmw_implementation/test/test_client.cpp:92]: @ahcorde what if the implementation identifier of either node or client does not match?

Comment (ahcorde) [reply test_rmw_implementation/test/test_client.cpp:92]: Related https://github.com/ros2/rmw/pull/280

Comment (ahcorde) [reply test_rmw_implementation/test/test_client.cpp:92]: related: 
 - https://github.com/ros2/rmw_connext/pull/466
 - https://github.com/ros2/rmw_fastrtps/pull/451

**Comment by hidmic** [comment]
Alright, PR rebased on top of #138.

**Comment by hidmic** [comment]
CI up to `test_rmw_implementation` and `rcl`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12522/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7491/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10234/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12441/)

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Review by Lobotuerk — APPROVED** [review]
LGTM

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by Lobotuerk** [comment]
Depends on https://github.com/ros2/rosidl_typesupport_connext/pull/60 to be released and rolling to be upped

**Comment by ahcorde** [comment]
@ros-pull-request-builder retest this please.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by hidmic** [comment]
It passed ! Finally all green. Merging!

**Pull request merged**